stdlib, configs: Add RiscvDemoBoard

erin-le opened the pull request:
This PR adds a RiscvDemoBoard that can be used with both SE and FS mode.This was tested using the workloads riscv-matrix-multiply-run for SE and riscv-ubuntu-20.04-boot for FS. Two example config scripts have also been added.

erin-le committed:
stdlib: Add RiscvDemoBoard

This commit adds a RiscvDemoBoard that can be used with both SE and
FS mode. The commented out parts of the board configuration are from
the X86DemoBoard while the uncommented parts are from
configs/example/gem5_library/riscv-ubuntu-run.py. This was tested
using the workloads riscv-matrix-multiply-run for SE and
riscv-ubuntu-20.04-boot for FS.

Change-Id: I5d338562f92c27a89101d6a8e70090e91faafe6f

erin-le committed:
stdlib: Remove comments from RiscvDemoBoard

Change-Id: Ic8578825051b83903de89d5408a6949c906edac5

erin-le committed:
configs: add example configuration scripts for RiscvDemoBoard

Change-Id: I645e520b390654cae7fb7ca5aa0f28ad83cfbd4a

erin-le committed:
configs: Add example script for Ubuntu 24.04 and RiscvDemoBoard

Change-Id: I3844390f1b84fa0e6d795bac3489dba6b5cbf543

Harshil2107 commented on configs/example/gem5_library/riscv-demo-board-se.py:
```diff
@@ -0,0 +1,66 @@
+board = RiscvDemoBoard(is_fs=False)
+
+# Here we set a workload.
+board.set_workload(obtain_resource(args.workload, args.version))
```
May be making the version an optional argument would be better, as if the use wants the latest version they can just pass the id.

erin-le committed:
configs: clarify args to obtain_resource in riscv-demo-board-se

Change-Id: If2b35c426e078edf4985d856bf78ec0af96df3ff

erin-le committed:
stdlib: change numbers in RiscvDemoBoard

This commit changes the cache/memory sizes and clock frequency
of the RiscvDemoBoard to better reflect real values and
be more similar to the ArmDemoBoard.

Change-Id: Ibe2ecc05cd2d6afaf446cdd6f9e28ba9add8b416

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-se.py:
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2021 The Regents of the University of California
```
Update the copyright here to 2024

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run-24.04.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2021 The Regents of the University of California
```
Update the copyright year

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run.py:
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2021 The Regents of the University of California
```
Update the copyright year

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run-24.04.py:
```diff
@@ -0,0 +1,102 @@
+    m5.stats.dump()
+    print("Reset stats at the end of the ROI!")
+    m5.stats.reset()
+    yield False
```
Isn't the `workend` exit event event the last exit event where we exit to the Python script? If true, shouldn't this be `yield True`?

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run-24.04.py:
```diff
@@ -0,0 +1,102 @@
+    print("Dump stats at the end of the ROI!")
+
+    m5.stats.dump()
+    print("Reset stats at the end of the ROI!")
+    m5.stats.reset()
```
Why do we reset stats at the end or the ROI? Isn't this the end of the simulation?

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run-24.04.py:
```diff
@@ -0,0 +1,102 @@
+def handle_workbegin():
+    print("Done booting Linux")
+    print("Dumping stats at beginning of ROI!")
+    m5.stats.dump()
```
Why do we want to dump stats here? If we're at the start of the workbegin then isn't  all the stats before just useless OS stuff?

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-ubuntu-run-24.04.py:
```diff
@@ -0,0 +1,102 @@
+    yield False
+
+
+def exit_event_handler():
+    print("first exit event: Kernel booted")
+    yield False
+    print("second exit event: In after boot")
+    yield False
+    print("third exit event: After run script")
+    print("Dump stats at end of workload!")
+    yield True
```
So there's a final "Exit" type exit event after the workend has been called? I'm very confused.

BobbyRBruce commented on src/python/gem5/prebuilt/demo/riscv_demo_board.py:
```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2021 The Regents of the University of California
```
Copyright year needs fixed here.

BobbyRBruce commented on configs/example/gem5_library/riscv-demo-board-se.py:
```diff
@@ -0,0 +1,69 @@
+
+# Here we set up the board. The prebuilt RiscvDemoBoard allows for FS
+# or SE simulation.
+board = RiscvDemoBoard(is_fs=False)
+
+# Here we set a workload. If nothing is passed for resource_version, the
+# latest resource version compatible with the gem5 version will be obtained.
+board.set_workload(
+    obtain_resource(resource_id=args.workload, resource_version=args.version)
+)
```
why do you need to declare `is_fs` in the `RiscvDemoBoard` constructor? Can't this just be infered by what workload is passed? I.e., if a FS is one is called then setup teh FS board, if not, setup the SE workload.

If you did that you could just merge these three demos board script together into one

erin-le committed:
stdlib, configs: infer fs or se in RiscvDemoBoard via workload

This commit modifies the RiscvDemoBoard so that it sets whether it
is in FS or SE mode based on the workload instead of requiring the
user to pass it to the constructor. This commit also consolidates
the RiscvDemoBoard configuration scripts into one script.

Change-Id: I524d73106d9641553c5393b5d9ee6e810b750310

erin-le committed:
configs: modify riscv demo script to only dump ROI statso

Change-Id: I5336b6db6b613e3adf8be0b6e1dfdb0853079240

erin-le committed:
stdlib: fix copyright year; modify riscv_demo_board.py

This commit fixes the copyright year in riscv_demo_board.py and
also modifies get_io_bus and get_mem_side_coherent_io_port to use
has_io_bus and has_coherent_io.

Change-Id: I18fb96059da1e05a19a43eb65e0f0cded1a59c67

BobbyRBruce commented:
I can't push to this PR but here's the patch that fixes this:

```patch
index 555f723df1..f5c536599e 100644
--- a/src/python/gem5/components/boards/riscv_board.py
+++ b/src/python/gem5/components/boards/riscv_board.py
@@ -26,7 +26,10 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import os
-from typing import List
+from typing import (
+    List,
+    Optional,
+)
 
 import m5
 from m5.objects import (
@@ -498,16 +501,26 @@ class RiscvBoard(AbstractSystemBoard, KernelDiskWorkload):
         return "/dev/vda"
 
     @overrides(AbstractSystemBoard)
-    def _pre_instantiate(self):
-        if len(self._bootloader) > 0:
-            self.workload.bootloader_addr = 0x0
-            self.workload.bootloader_filename = self._bootloader[0]
-            self.workload.kernel_addr = 0x80200000
-            self.workload.entry_point = 0x80000000  # Bootloader starting point
-        else:
-            self.workload.kernel_addr = 0x0
-            self.workload.entry_point = 0x80000000
-        self._connect_things()
+    def _pre_instantiate(self, full_system: Optional[bool] = None):
+        # This is a bit of a hack necessary to get the RiscDemoBoard working
+        # At the time of writing the RiscvBoard does not support SE mode so
+        # this branch looks pointless. However, the RiscvDemoBoard does and
+        # needs this logic in place.
+        #
+        # This should be refactored in the future as part of a chance to have
+        # all boards support both FS and SE modes.
+        if self._is_fs:
+            if len(self._bootloader) > 0:
+                self.workload.bootloader_addr = 0x0
+                self.workload.bootloader_filename = self._bootloader[0]
+                self.workload.kernel_addr = 0x80200000
+                self.workload.entry_point = (
+                    0x80000000  # Bootloader starting point
+                )
+            else:
+                self.workload.kernel_addr = 0x0
+                self.workload.entry_point = 0x80000000
+        super()._pre_instantiate(full_system=full_system)
 
     @overrides(KernelDiskWorkload)
     def _add_disk_to_board(self, disk_image: AbstractResource):
diff --git a/src/python/gem5/prebuilt/demo/riscv_demo_board.py b/src/python/gem5/prebuilt/demo/riscv_demo_board.py
index e4ef627895..cd73e8efe6 100644
--- a/src/python/gem5/prebuilt/demo/riscv_demo_board.py
+++ b/src/python/gem5/prebuilt/demo/riscv_demo_board.py
@@ -107,24 +107,6 @@ class RiscvDemoBoard(RiscvBoard, SEBinaryWorkload):
             cache_hierarchy=cache_hierarchy,
         )
 
-    # Taken from Riscv Matched board. Below are functions that are needed to get
-    # SE mode to work.
-    @overrides(RiscvBoard)
-    def _pre_instantiate(self):
-        if self._is_fs:
-            if len(self._bootloader) > 0:
-                self.workload.bootloader_addr = 0x0
-                self.workload.bootloader_filename = self._bootloader[0]
-                self.workload.kernel_addr = 0x80200000
-                self.workload.entry_point = (
-                    0x80000000  # Bootloader starting point
-                )
-            else:
-                self.workload.kernel_addr = 0x0
-                self.workload.entry_point = 0x80000000
-
-        self._connect_things()
-
     @overrides(RiscvBoard)
     def _setup_board(self) -> None:
         if self._is_fs:
```

BobbyRBruce commented:
The commit message I was going to have going with this patch (if it helps):

```commit
 stdlib,arch-riscv: Fix RISCV DemoBoard
    
    1. RiscvBoard did not have the correct signature for _pre_instantiate
       (missing `full_system` param).
    2. RiscvDemoBoard required some `is_fs` logic moved down to the
       RiscvBoard.
```

erin-le committed:
Merge branch 'develop' into riscv-demo-board-develop

erin-le committed:
stdlib, arch-riscv: add BobbyRBruce's changes for RVDemoBoard

This commit adds BobbyRBruce's changes to the RiscvDemoBoard to
make it work with the changes to pre-instantiate. The rest of
his commit message is below:
    1. RiscvBoard did not have the correct signature for _pre_instantiate
       (missing `full_system` param).
    2. RiscvDemoBoard required some `is_fs` logic moved down to the
       RiscvBoard.

Change-Id: Ie35e1bc0d1719eafb0ba53c3838fecefe85daae8

erin-le committed:
stdlib, arch-riscv: undo accidental deletion of line

powerjg commented on src/python/gem5/components/boards/riscv_board.py:
```diff
@@ -502,14 +502,24 @@ def get_disk_device(self):
-        else:
-            self.workload.kernel_addr = 0x0
-            self.workload.entry_point = 0x80000000
+        # This is a bit of a hack necessary to get the RiscDemoBoard working
+        # At the time of writing the RiscvBoard does not support SE mode so
+        # this branch looks pointless. However, the RiscvDemoBoard does and
+        # needs this logic in place.
+        #
+        # This should be refactored in the future as part of a chance to have
+        # all boards support both FS and SE modes.
```
This is OK, but couldn't you override this in `RiscvDemoBoard` and have something like:

```python
def _pre_instantiate(self...):
     if self._is_fs:
        super()._pre_instantiate()
    else:
        self.workload.kernel_addr = 0x0
...
        AbstractBoard._pre_instantiate(self, full_system)
```
```

BobbyRBruce replied:
The problem i have with that is you'd need to copy and paste the `AbstractBoard`'s `_pre_instantiate` function into the `else` block which is a bit annoying. In short, for FS mode we want to run the `RiscvBoard`'s `_pre_instantiate` and for SE mide we just want to run the `AbstractBoard`'s. I tried `super().super()._pre_instantiate()` but Python didn't like it. Perhaps there's a way to do it but in the end i went with the solution I had working.

powerjg replied:
I see. I *could* do `AbstractBoard._pre_instantiate(self, ...)` in the `else`, but that isn't ideal either. Feel free to ignore this comment.

erin-le commented:
I tested this change using the workloads riscv-matrix-multiply-run and riscv-ubuntu-24.04-boot. Both simulations run, so it looks like this fixed the problems that the RiscvDemoBoard was having with the pre-instantiate change.
